### PR TITLE
doc: multi-image: add build file table

### PR DESCRIPTION
**Pull request by greg-fer**
Added build file table to explain the purpose of different files.
KRKNWK-13103.

Signed-off-by: Grzegorz Ferenc <Grzegorz.Ferenc@nordicsemi.no>

----

- [x] Squash to one commit

### CODE REVIEW

**Comment by greg-fer** [comment]
@trond-snekvik , @hakonfam , @tejlmand , I'd appreciate any feedback, also because the current draft includes table copied from ug_thingy53, so certainly not comprehensive of all cases, albeit with some minor edits.

File: doc/nrf/ug_multi_image.rst
Comment (trond-snekvik) [inline doc/nrf/ug_multi_image.rst:85]: In this case, I think "HEX file", "hex file" or "Intel hex file" is appropriate, as it's a valid concept on its own, and no longer an abbreviation: https://en.wikipedia.org/wiki/Intel_HEX

**Review by trond-snekvik — COMMENTED** [review]
One comment on nomenclature, but the tables look good to me :+1:

**Review by trond-snekvik — APPROVED** [review]

File: doc/nrf/gs_programming.rst
Comment (hakonfam) [inline doc/nrf/gs_programming.rst:173]: Do we really want to explain what an elf file is? It is a standard concept, and we should simply state that it is the elf file for the image being built.

File: doc/nrf/gs_programming.rst
Comment (hakonfam) [inline doc/nrf/gs_programming.rst:175]: We should also state that this is the file that would be transferred in an actual update procedure, as opposed to the hex files we use when we use nrfjprog to emulate an update procedure.

**Review by hakonfam — CHANGES_REQUESTED** [review]

File: doc/nrf/gs_programming.rst
Comment (greg-fer) [reply doc/nrf/gs_programming.rst:155]: What do you mean with "merging all zephyr.hex"?

**Comment by greg-fer** [comment]
@hakonfam , edits applied, please check again.

File: doc/nrf/gs_programming.rst
Comment (hakonfam) [reply doc/nrf/gs_programming.rst]: Each image in a multi image build will have a `zephyr.hex` associated with it. We merge (that is, combine them into a single file) all of these into `merged.hex`

**Review by hakonfam — APPROVED** [review]

**Review by kkasperczyk-no — APPROVED** [review]

File: doc/nrf/gs_programming.rst
Comment (Damian-Nordic) [inline doc/nrf/gs_programming.rst:133]:
```suggestion
The symbol table exists as metadata in a :file:`.elf` or :file:`.exe` file and is not included when the image is converted to a hexadecimal file for programming.
```

File: doc/nrf/gs_programming.rst
Comment (Damian-Nordic) [inline doc/nrf/gs_programming.rst:155]: Not used by Nordic Semiconductor's build targets? Did you mean "Not used by single core build targets"?

File: doc/nrf/gs_programming.rst
Comment (Damian-Nordic) [reply doc/nrf/gs_programming.rst:133]: Or `:file:'.hex'`

File: doc/nrf/gs_programming.rst
Comment (Damian-Nordic) [inline doc/nrf/gs_programming.rst:136]: I'm not sure what these two lines try to convey. Can anyone help? Unless I misunderstand something this seems to be pretty technical information that is either known for users or it cannot be explained in two sentences, so probably can be removed.

**Review by Damian-Nordic — APPROVED** [review]
Found some nits, but approving (in particular, the Matter changes).

File: doc/nrf/gs_programming.rst
Comment (greg-fer) [reply doc/nrf/gs_programming.rst:155]: From what I understood, it's not used at all, but rather used to create merged.hex... Is that incorrect?

File: doc/nrf/gs_programming.rst
Comment (greg-fer) [reply doc/nrf/gs_programming.rst:136]: I wasn't sure if we need this when I was moving these lines from below. Looks like an introduction to why we can have multi-image builds.

File: doc/nrf/gs_programming.rst
Comment (Damian-Nordic) [reply doc/nrf/gs_programming.rst:155]: But it is a description of `merged.hex`. I can tell that if you build a multi-image application for nRF52840 it is used when you run `west flash`.

File: doc/nrf/gs_programming.rst
Comment (Damian-Nordic) [reply doc/nrf/gs_programming.rst:136]: Looks like the original author (Torsten) is absent, so we can leave it as it is, but reading this for the first time I find it rather confusing and not crucial to grasp the idea of multi-image builds.

File: doc/nrf/gs_programming.rst
Comment (greg-fer) [inline doc/nrf/gs_programming.rst:174]: Add for Zigbee after talking to @sebastiandraus :

:file:`<file_name>.zigbee` | :ref:`ug_zigbee`-specific OTA image that contains the Zigbee application with the Zigbee OTA header used for providing information about the image to the OTA server. The *<file_name>* includes manufacturer's code, image type, file version, optional comment, and the sample name. For example, :file:`127F-0141-01020003-light_switch.zigbee`. | DFU over Zigbee for both single-core and multi-core applications in the |NCS| v2.0.0 and later.

File: doc/nrf/gs_programming.rst
Comment (greg-fer) [reply doc/nrf/gs_programming.rst:155]: I meant, merged_domain.hex, sorry...

Comment (greg-fer) [reply doc/nrf/gs_programming.rst]: Should be clear now, @Damian-Nordic , after some clarifications from Haakon...

**Comment by greg-fer** [comment]
@sebastiandraus , please check and approve if all looks OK for you...

**Review by Damian-Nordic — APPROVED** [review]

**Comment by greg-fer** [comment]
Waiting for feedback from @tejlmand before final rebase and merge request...

**Review by LuDuda — APPROVED** [review]
LGTM :+1: Appreciate it!

File: doc/nrf/gs_programming.rst
Comment (tejlmand) [reply doc/nrf/gs_programming.rst:136]: @Damian-Nordic i'm not the author of this. I'm the the committer (because I merged the PR), the authors are:
@kyberdin and @FrancescoSer 
Ref: https://github.com/nrfconnect/sdk-nrf/commit/2f76e95ad435161135111e16b00f2d0eb04338b3

Comment (kyberdin) [reply doc/nrf/gs_programming.rst:136]: Agreed with @Damian-Nordic, this is unnecessary information for the user and probably just leads to more confusion. The original docs were worded poorly to begin with and this doesn't convey the actual informative intent, but users that care about that level of detail probably already know it anyway.

This line and Line 137 are probably just better off removed to keep the info here simpler and more cohesive.

File: doc/nrf/gs_programming.rst
Comment (tejlmand) [inline doc/nrf/gs_programming.rst:137]: Instead of what ? and when does it do it ?
How about rewriting the the lines into something like this:
```suggestion
As one of the last build steps, the linker will process all object files by locating code, data and symbols in sections in the final ELF file.
The linker replaces all symbol references to code and data with addresses.
A symbol table is created which maps addresses to symbol names, which is used by debuggers.

When an ELF file is converted into another format, such as HEX or binary, the symbol table is lost.
```
I left out some details in the above (the real world is a little more complex), but tried be accurate in as few words as possible.

File: doc/nrf/gs_programming.rst
Comment (tejlmand) [inline doc/nrf/gs_programming.rst:140]: > count the number of times the linker runs.

How can I judge the number of images by the number of linker runs ?
Zephyr build system may create intermediate binaries, which means that when creating a final build, the linker may run 1, 2, or 3 times just for a single image.

File: doc/nrf/gs_programming.rst
Comment (tejlmand) [inline doc/nrf/gs_programming.rst:205]: I wouldn't call an ELF file a meta data file.

It's the most complete file generated by the system, and only because our flash tools doesn't support flashing of ELF files directly it must be converted. (Plus the fact that we might manipulate / merge hex files, but that's a different story)
Some tools actually supports flashing of ELF files.

**Review by tejlmand — CHANGES_REQUESTED** [review]
Thanks for this, looks much better.

A couple of comments / suggestion in areas where I believe we can be more precise / clearer on what we want to describe to the reader.

File: doc/nrf/gs_programming.rst
Comment (tejlmand) [reply doc/nrf/gs_programming.rst:140]: https://docs.zephyrproject.org/latest/build/cmake/index.html

**Comment by greg-fer** [comment]
Applied edits (and moved the new sections elsewhere). @tejlmand , @kyberdin , please check one more time when you have time.

**Review by kyberdin — APPROVED** [review]

**Review by hakonfam — APPROVED** [review]

**Review by tejlmand — APPROVED** [review]
lgtm. :+1:

**Review by FrancescoSer — APPROVED** [review]

**Pull request merged**
